src/cargo/ops/cargo_compile/mod.rs

```diff
@@ -476,6 +481,88 @@ pub fn create_bcx<'a, 'gctx>(
             .extend(args);
     }
 
+    if let Some(check_cfg) = target_data.check_cfg() {
+        if check_cfg.exhaustive {
+            let mut error_count = 0;
+            let mut max_lint_level = LintLevel::Allow;
+
+            let target_cfgs = gctx.target_cfgs()?;
+            for (key, _fields) in target_cfgs {
+                if !key.starts_with("cfg(") || !key.ends_with(')') {
+                    continue;
+                }
+
+                let cfg_expr = &key[4..key.len() - 1];
+                let Ok(cfg_expr) = CfgExpr::from_str(cfg_expr) else {
+                    continue;
+                };
+
+                warn_on_unexpected_cfgs(
+                    gctx,
+                    check_cfg,
+                    &cfg_expr,
+                    LintLevel::Warn,
+                    &mut error_count,
+                    None,
+                    "",
+                )?;
+            }
+
+            for unit in &units {
```

Comment:
Another problem with doing this in compilation is that a package can have multiple build units and they share dependencies, so you can get duplicate dependency reports